docs: Add more documentation (#18)
aneoconsulting/ArmoniK.Extensions.Cpp: .docs/content/3.end2end.md

- The invocation handler to return a handle response or error.

Before submitting tasks to ArmoniK, we must first provide the implementation of an abstract class ```IServiceInvocationHandler``` for the invocation handler. This class contains two pure virtual methods which are ```HandleResponse``` and ```HandleError``` that must be implemented.
- The HandleResponse is the callback function to manage the result in case of success. The arguments of this function are serialized data from the server as a result of the method invoked by the client and the task id that invoked the callback.
- The HandleError is the callback function to retrieve errors or exceptions. The arguments are the exception sent to the control plane and the task id that invoked the callback.

The ```WaitResults```retrieves of submitted tasks. This method waits for a given set of tasks to retrieve results. If no tasks are specified, then the method waits for all the tasks launched with this service to be completed. If a set of tasks is specified then the waitBehavior parameter defines when the function returns. This method takes as arguments a set of task IDs, a waitBehavior, and a waitOptions.

> - Task ids the ids to wait the results for. If other tasks have been launched with the service, they may be handled at the same time as the tasks in the set. If empty, the function will wait on the results of all the tasks submitted with this service.

Reviewer: ```suggestion
- Task ids is an array set with the ids of tasks for which we are waiting for the results. If other tasks have been launched with the service, they may be handled at the same time as the tasks in the set. If empty, the function will wait on the results of all the tasks submitted with this service.
```

- WaitBahavior specifies if it should wait for all tasks, any task completion, and/or stop waiting if a result is aborted from the given task_ids list. If the task_ids set is empty, All and Any are synonymous. If the waitBehavior contains the BreakOnError flag, the function exits after a result sent by this service is aborted.
- WaitOptions specifies the amount of time between result status checks.

### Configure client

The constructor of the SessionService class takes as arguments the properties and the logger. The properties defined in ```ArmoniK::Sdk::Common``` namespace contain the configuration object and the session task options.

> - The configuration object consists of a JSON file or environment variables that define the following:

Reviewer: ```suggestion
- The configuration object is loaded from a JSON file or from environment variables that define the following:
```

>   - Endpoint: The control plane address to connect.

Reviewer: ```suggestion
  - Endpoint: The control plane address to which clients connect.
```

  - mTLS: enable or disable mTLS.
  - caCert: The Server certificate file to validate mTLS.
  - clientCert: The client certificate fil in a pem format.
  - clientKey: The client key file in a pem format.
  - sslValidation: Disable the ssl strong validation of ssl certificate (default: enable => true).

> The environment variables override the configuration in the JSON file.s

Reviewer: ```suggestion
The environment variables override the configuration in the JSON files.
```

- The session task options contain information required by a session such as the following:
  - MaxRetries (default value is set to 2)
  - Priority (default value is set to 1)
  - AppName the name of the worker shared library
  - AppVersion the version of the library (can be empty)
  - AppNamespace the service namespace
  - AppService the service name

## Worker

> The ArmoniK dynamic worker deployed with the infrastructure is responsible for receiving tasks from clients and executing them based on the task options and the task payloads. The methods to be invoked are determined in the task payload with their serialized arguments. The implementation of those methods, as long as the five mandatory symbols to use ArmoniK, are in the worker-shared library as a ServiceBase. In the following, we show how to set up and build this shared library.

Reviewer: ```suggestion
The ArmoniK dynamic worker deployed with the infrastructure is responsible for receiving tasks from clients and executing them based on the task options and the task payloads. The methods to be invoked are determined in the task payload with their serialized arguments. The implementation of those methods must be present in the worker-shared library as a ServiceBase object. In the following, we show how to set up and build this shared library.
```

Reviewer: it is too early to talk about five mandatory symbols.

### Setup worker shared lib

One can set up a shared library in order to use the ArmoniK Worker. To do so, an implementation of a service class inheriting of ArmoniK::Sdk::Worker::ServiceBase must be provided. An example of this kind of implementation can be found in [Echo service](https://github.com/aneoconsulting/ArmoniK.Extensions.Cpp/blob/main/ArmoniK.SDK.Worker.Test/include/EchoService.h) or [Addition service](https://github.com/aneoconsulting/ArmoniK.Extensions.Cpp/blob/main/ArmoniK.SDK.Worker.Test/include/AdditionService.h). The examples provide the service class implementing the methods called when entering, calling or leaving a session. When using this approach, one may only provide the implementation of the creation of ```armonik_create_service``` as in [here](https://github.com/aneoconsulting/ArmoniK.Extensions.Cpp/blob/main/ArmoniK.SDK.Worker.Test/src/ServiceDispatch.cpp)

> It is also possible to set up the shared library without inheriting ArmoniK::Sdk::Worker::ServiceBase. In that scenario, one should provide the implementation of all the five mandatory symbols:

Reviewer: ```suggestion
It is also possible to set up the shared library without inheriting ArmoniK::Sdk::Worker::ServiceBase. In that scenario, one should provide the implementation of the following mandatory methods:
```

Reviewer: should add details when each method is called by the worker runtime.
